reference/Tensor.cpp

```diff
@@ -0,0 +1,365 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+   Copyright 2022 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "Tensor.h"
+
+#include <algorithm>
+#include <complex>
+
+#include "llvm/ADT/APFloat.h"
+#include "mlir/AsmParser/AsmParser.h"
+#include "mlir/IR/BuiltinTypes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+namespace {
+
+int64_t getSizeInBytes(Type type) {
+  if (auto shapedType = type.dyn_cast<ShapedType>())
+    return shapedType.getNumElements() *
+           getSizeInBytes(shapedType.getElementType());
+
+  if (type.isIntOrFloat())
+    return std::max(type.getIntOrFloatBitWidth(), (unsigned)8) / 8;
+
+  if (auto complexType = type.dyn_cast<mlir::ComplexType>())
+    return getSizeInBytes(complexType.getElementType()) * 2;
+
+  llvm_unreachable("Unsupported element type");
+}
+
+}  // namespace
+
+namespace detail {
+
+Buffer::Buffer(ShapedType type) : type_(type), data_(getSizeInBytes(type), 0) {}
+
+Buffer::Buffer(ShapedType type, void *data)
+    : Buffer(type, static_cast<const void *>(data)) {}
+
+Buffer::Buffer(ShapedType type, const void *data)
+    : type_(type),
+      data_(static_cast<const char *>(data),
+            static_cast<const char *>(data) + getSizeInBytes(type)) {}
+
+}  // namespace detail
+
+Tensor::Tensor() {}
+
+Tensor::Tensor(ShapedType type)
+    : impl_(llvm::makeIntrusiveRefCnt<detail::Buffer>(type)) {}
+
+Tensor::Tensor(DenseElementsAttr attr) {
+  // Note: Soon we will be deprecating the use of
```

Comment:
How about providing more context here, e.g.: `TODO(sdasgup3): We're using DenseElementsAttr::getRawData() here for simplicity, because it provides a contiguous representation of underlying data in most cases. However, this doesn't always work (e.g. for splat or for i1), so we'll be migrating to something more reliable in the near future`?


 but its return value is not always the expected byte array, atleast for splat and boolean values